tls/s2n_kem.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "tls/s2n_kem.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+int s2n_kem_generate_keypair(struct s2n_kem_keypair *kem_keys)
+{
+    const struct s2n_kem *kem = kem_keys->negotiated_kem;
+    notnull_check(kem->generate_keypair);
+
+    eq_check(kem_keys->public_key.size, kem->public_key_length);
+    notnull_check(kem_keys->public_key.data);
+
+    // The private key is needed for client_key_recv and must be saved
+    GUARD(s2n_alloc(&kem_keys->private_key, kem->private_key_length));
+
+    GUARD(kem->generate_keypair(kem_keys->public_key.data, kem_keys->private_key.data));
+    return 0;
+}
+
+int s2n_kem_encapsulate(const struct s2n_kem_keypair *kem_keys, struct s2n_blob *shared_secret,
+                        struct s2n_blob *ciphertext)
+{
+    const struct s2n_kem *kem = kem_keys->negotiated_kem;
+    notnull_check(kem->encapsulate);
+
+    eq_check(kem_keys->public_key.size, kem->public_key_length);
+    notnull_check(kem_keys->public_key.data);
+
+    eq_check(ciphertext->size, kem->ciphertext_length);
+    notnull_check(ciphertext->data);
+
+    GUARD(s2n_alloc(shared_secret, kem->shared_secret_key_length));
+
+    GUARD(kem->encapsulate(ciphertext->data, shared_secret->data, kem_keys->public_key.data));
+    return 0;
+}
+
+int s2n_kem_decapsulate(const struct s2n_kem_keypair *kem_keys, struct s2n_blob *shared_secret,
+                        const struct s2n_blob *ciphertext)
+{
+    const struct s2n_kem *kem = kem_keys->negotiated_kem;
```

Comment:
what if kem_keys is null? seems weird to have a not null check for a member, but not the parent.

Reply:
Good catch, I fixed all of these.